S3 Backup design doc

mvandenburgh opened the pull request:
This PR lays out a design for S3 backup using S3 Replication and the Glacier Deep Storage class. Related #524

mvandenburgh committed:
Design doc for S3 backup

CodyCBakerPhD commented on doc/design/s3-backup.md:
```diff
@@ -0,0 +1,84 @@
+
+**Data Transfer Costs**:
+
+- Same-region data transfer between S3 buckets is free
```
The definition on line https://github.com/dandi/dandi-archive/pull/2627/files#diff-9a9b9bb814eeea229ca42a1f536abf895736f6aeb248def55cc122469f619d0fR6 indicates the backup will live in a separate region though?

mvandenburgh replied:
Ah thanks, I missed that - we originally considered a bucket in a separate region, but settled on same-region. I pushed a fix

CodyCBakerPhD commented on doc/design/s3-backup.md:
```diff
@@ -0,0 +1,84 @@
+
+- Glacier Deep Archive retrieval:
+    - $0.02/GB (standard, 12-hour retrieval)
+    - $0.025/GB (bulk retrieval, can take up to 48 hours)
```
Can you produce worst-case full bucket retrieval summary estimates here? Similar to https://github.com/dandi/dandi-archive/pull/2627/files#diff-9a9b9bb814eeea229ca42a1f536abf895736f6aeb248def55cc122469f619d0fR60

CodyCBakerPhD replied:
Since that cost is very important to consider to get full use from this endeavor

mvandenburgh committed:
Fix outdated info

We are no longer considering the use of a bucket in a different region.

waxlamp commented on doc/design/s3-backup.md:
```diff
@@ -0,0 +1,84 @@
+
+### Future Costs
+
+The DANDI Archive is expecting a ramp-up in data volume of 6 PB of new data over each of the next five years, culminating in a total of 30PB.
```
> I thought it was 1 PB per year for a total of 6 PB by the end? And then a separate potential project that was also several PB

I may have misunderstood. @kabilar probably knows the answer.

> Also we didn't get 100% confirmation of S3 bucket size extension yet either did we? (guessing it is fine for first 1-2 years up to 2-3 PB total storage)

I was not privy to those conversations. This one is for @satra.

waxlamp committed:
Update cost analysis with accurate inputs

We are expecting a bulk of 6PB over the next 5 years, not 30PB.

waxlamp committed:
Fix typo in bulk retrieval cost

waxlamp committed:
Add worst-case disaster recovery cost estimate

CodyCBakerPhD commented on doc/design/s3-backup.md:
```diff
@@ -0,0 +1,84 @@
+
+The DANDI Archive is expecting a ramp-up in data volume of 6 PB of new data over each of the next five years, culminating in a total of 30PB.
+
+Scaling up the previous analysis means that the monthly costs will be projected to rise to a total of **~$31,000/month** once all of that data is seated. While $1000/month may be a feasible ongoing cost, $30000/month is not.
```
The output of `df -hT .`

```
Filesystem                   Type
hstor004-n1:/group/dandi/001    nfs4
```

CodyCBakerPhD commented on doc/design/s3-backup.md:
```diff
@@ -0,0 +1,90 @@
+1. **S3 Replication** to copy objects from the primary bucket to a backup bucket.
+2. **Delete Marker Replication** to propagate soft deletions to the backup.
+3. **Bucket Versioning** on both buckets to create a history-based backup and support the replication and recovery mechanisms
+4. **Optional Lifecycle Policies** on the backup bucket to permanently delete objects after a retention period
```
```suggestion
3. **Bucket Versioning** on both buckets to create a history-based backup and support the replication and recovery mechanisms.
4. **Optional Lifecycle Policies** on the backup bucket to permanently delete objects after a retention period.
```

for consistency (the above lines end in periods)

CodyCBakerPhD commented on doc/design/s3-backup.md:
```diff
@@ -0,0 +1,90 @@
+
+$$
+6\ \rm{PB} \times 1024\ \rm{TB}/\rm{PB} \times 1024\ \rm{GB}/\rm{TB} \times \$0.021/GB/mo \approxeq \$130000/mo
+$$
```
Recommend adjusting as a I showed above

waxlamp replied:
6b69903917d539ef3aa17558b31a16f5788d4dbf

waxlamp committed:
Add missing periods

Co-authored-by: Cody Baker <51133164+CodyCBakerPhD@users.noreply.github.com>

waxlamp committed:
Improve readability of LaTeX and fix conversion units

Co-authored-by: Cody Baker <51133164+CodyCBakerPhD@users.noreply.github.com>

waxlamp committed:
Fix conversion units

When the AWS docs say "GB", they mean 10^9 bytes, not 2^30.

Co-authored-by: Cody Baker <51133164+CodyCBakerPhD@users.noreply.github.com>

waxlamp committed:
Fix dollar amount formatting

Co-authored-by: Cody Baker <51133164+CodyCBakerPhD@users.noreply.github.com>

waxlamp committed:
Improve readability and fix conversion factors

waxlamp committed:
Fix cost analysis

Clarify purpose of calculating the expected bucket storage cost covered
by AWS already.

waxlamp commented on doc/design/s3-backup.md:
```diff
@@ -0,0 +1,90 @@
+6\ \rm{PB} \times 1024\ \rm{TB}/\rm{PB} \times 1024\ \rm{GB}/\rm{TB} \times \$0.021/GB/mo \approxeq \$130000/mo
+$$
+
+while the associated backup costs would represent only $`\$31000 / \$660000 \approxeq 4.6\%`$ of the cost of the storage itself.
```
b3e0a9d

CodyCBakerPhD commented on doc/design/s3-backup.md:
```diff
@@ -0,0 +1,92 @@
+$$
+
+while the associated backup costs would represent only an additional $`\$5900 / \$126000 \approxeq 4.6\%`$ of the cost of the storage itself.
+To help provide a significant level of safety to an important dataset, AWS may be willing to cover such a low marginal cost.
```
```suggestion
To help provide a significant level of safety to an important database, it may be worth reaching out to see if AWS may be willing to cover such a low marginal cost.
```

Original wording sounds as if we are speaking for AWS

CodyCBakerPhD replied:
Please remove this line as it has been confirmed that the Open Data program will not cover backup.

yarikoptic replied:
```suggestion
```

kabilar commented on doc/design/s3-backup.md:
```diff
@@ -0,0 +1,92 @@
+$$
+
+while the associated backup costs would represent only an additional $`\$5900 / \$126000 \approxeq 4.6\%`$ of the cost of the storage itself.
+To help provide a significant level of safety to an important dataset, AWS may be willing to cover such a low marginal cost.
```
@satra The design here looks good, but the main open question and blocker is cost.

Would the AWS Open Data program be open to covering the back up storage costs as well (~$12,000/year for a 1 PB backup, and ~$72,000/year for the projected 6 PB backup)? See the doc for more details.

CodyCBakerPhD committed:
Merge pull request #2829 from CodyCBakerPhD/patch-1

Update status for initial s3 backup doc
